# Evaluating The Performance of Various Bayesian Models

In [1]:
import fourtenmodule
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.feature_selection import VarianceThreshold


## Synthetic Data

### Initial Set-Up

In [2]:
np.random.seed(410)
n=50
p = 5
# Create alternating columns of [0, 1] and [1, 0] in 3 columns to reduce experimental randomness
X = np.ones((n, p))
for i in range(0, X.shape[1], 2):
    X[i::2, i] = 0
    X[i+1::2, i] = 1

# Create true_beta based on logspace
true_beta = np.logspace(0, 1, p, base=2)
sigma_true = 1
print(true_beta)

# Generate y with added random noise
y = np.dot(X, true_beta) + stats.norm(0,sigma_true).rvs(n)

fourtenmodule.run_and_plot_models(X, y, true_beta, n_components=2)

[1.         1.18920712 1.41421356 1.68179283 2.        ]

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 1 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


  → Passed in 2.84s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 1 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.89s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 1 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.14s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 1 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details


  → Passed in 1.37s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 82.229
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.27s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 0 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.85s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 100 tune and 100 draw iterations (400 + 400 draws total) took 0 seconds.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.73s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 7 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 14 seconds.
There were 254 divergences after tuning. Increase `target_accept` or reparameterize.
The rhat statistic is larger than 1.01 for some parameters. This indicates problems during sampling. See https://arxiv.org/abs/1903.08008 for details
The effective sample size per chain is smaller than 100 for some parameters.  A higher number is needed for reliable rhat and ess computation. See https://arxiv.org/abs/1903.08008 for details
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 46 seconds.
There were 1 divergences after tuning. Increase `target_accept` or reparameterize.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 10 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 82.562
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/sklearn/decomposition/_fastica.py:629: RuntimeWarning: invalid value encountered in divide
  K = (u / d).T[:n_components]  # see (6.33) p.140
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 0.8930     β0:0.782  β1:1.140  β2:0.849  β3:1.839  β4:2.777
Bayesian Robust Regression     0.8954     β0:0.817  β1:1.253  β2:1.211  β3:1.746  β4:2.363
Bayesian Linear Regression     0.8955     β0:0.817  β1:1.248  β2:1.227  β3:1.743  β4:2.360
Bayesian Variational Inference 0.9003     β0:0.963  β1:1.251  β2:1.382  β3:1.726  β4:2.055
Bayesian Ridge Regression      0.9020     β0:0.982  β1:1.217  β2:1.404  β3:1.719  β4:2.042
Bayesian ICR                   4.0547     β0:0.292  β1:0.000  β2:0.919  β3:0.000  β4:1.518
Bayesian PCR                   4.2911     β0:0.755  β1:0.000  β2:0.800  β3:0.000  β4:0.823


{'passing_models': {'Bayesian Linear Regression': <function fourtenmodule.bayesian_regression_mcmc(X, y, true_beta=None, draws=4000)>,
  'Bayesian Ridge Regression': <function fourtenmodule.bayesian_ridge_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Lasso': <function fourtenmodule.bayesian_lasso(X, y, true_beta=None, n_folds=5, draws=4000)>,
  'Bayesian Robust Regression': <function fourtenmodule.bayesian_robust_regression(X, y, true_beta=None, draws=4000)>,
  'Bayesian Variational Inference': <function fourtenmodule.bayesian_regression_vi(X, y, true_beta=None)>,
  'Bayesian PCR': <function fourtenmodule.bayesian_pcr(X, y, true_beta=None, n_components=None, draws=4000)>,
  'Bayesian ICR': <function fourtenmodule.bayesian_icr(X, y, true_beta=None, n_components=None, draws=4000)>},
 'results': [{'Model': 'Bayesian Linear Regression',
   'RMSE': np.float64(0.8954770744561744),
   'Beta Estimates': array([0.81743459, 1.24770735, 1.22711822, 1.74276605, 2.359622  ])},
  {'Mode

The true beta is given by [1, 1.18920712, 1.41421356, 1.68179283, 2]. The best model by RMSE appears to be the Bayesian Lasso Regression, with Linear and Robust regression producing similar scores. On the other hand, ICR and PCR produce significantly worse RMSE values due to the small number of predictors.


## Changing Parameters

### Sample Size

In [ ]:
# Define sample sizes to test
sample_sizes = [50, 100, 500, 1000, 1500]
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for n in sample_sizes:
    print(f"Running models for sample size: {n}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(sample_sizes, rmse, marker='o', label=model)
plt.xlabel("Sample Size")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Sample Size Increases")
plt.legend()
plt.grid()
plt.show()

In [ ]:
plt.figure(figsize=(8, 6))

for model, rmse in rmse_results.items():
    if model != "ICR":
        plt.plot(sample_sizes, rmse, marker='o', label=model)
        
plt.xlabel("Sample Size")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Sample Size Increases")
plt.legend()
plt.grid()
plt.show()

### Number of parameters

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases")
plt.legend()
plt.grid()
plt.show()

### Using highly sparse data

In [ ]:
# Define sample sizes to test
p_numbers = [5, 20, 30, 60, 100]
n=20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR" : [], "ICR": []}

for p in p_numbers:
    print(f"Running models for this many parameters: {p}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(p = p, high_corr = True)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of sample size
plt.figure(figsize=(8, 6))
for model, rmse in rmse_results.items():
    plt.plot(p_numbers, rmse, marker='o', label=model)
plt.xlabel("Number of Predictors")
plt.ylabel("RMSE of Estimated Beta")
plt.title("Model Performance as Number of Predictors Increases With Sparse True Signals")
plt.legend()
plt.grid()
plt.show()

### Magnitudes

#### Beta 0

In [ ]:
# Define different values where only one coefficient increases at a time
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

beta_variations = [[2**i, *true_beta[1:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "2", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 0 Variations")
plt.ylabel("RMSE of Estimated Beta 0")
plt.title("Model Performance as Beta 0 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


#### Beta 1

In [ ]:
p=5
# Fixed 5-element true_beta array
true_beta = np.logspace(0, 1, p, base=2)

# Define different values where only one coefficient increases at a time
beta_variations = [ [true_beta[0], 2**i, *true_beta[2:]] for i in range(5)]


n = 20
rmse_results = {"MCMC": [], "Ridge": [], "Lasso": [], "Robust": [], "VI": [], "PCR": [], "ICR": []}

for true_beta in beta_variations:
    print(f"Running models with true_beta: {true_beta}")
    rmse_mcmc, rmse_ridge, rmse_lasso, rmse_robust, rmse_vi, rmse_pcr, rmse_icr = fourtenmodule.run_models_and_evaluate(n = n, true_beta = true_beta)
    rmse_results["MCMC"].append(rmse_mcmc)
    rmse_results["Ridge"].append(rmse_ridge)
    rmse_results["Lasso"].append(rmse_lasso)
    rmse_results["Robust"].append(rmse_robust)
    rmse_results["VI"].append(rmse_vi)
    rmse_results["PCR"].append(rmse_pcr)
    rmse_results["ICR"].append(rmse_icr)

# Plot RMSE as a function of increasing individual beta values
plt.figure(figsize=(8, 6))
beta_labels = [
    "1", "1.41", "4", "8", "16"
]
for model, rmse in rmse_results.items():
    plt.plot(beta_labels, rmse, marker='o', label=model)
plt.xlabel("True Beta 1 Variations")
plt.ylabel("RMSE of Estimated Beta 1")
plt.title("Model Performance as Beta 1 Increases")
plt.xticks(rotation=45)
plt.legend()
plt.grid()
plt.show()


# Real Data

## Housing Data

In [2]:
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/housing.csv"
        return pd.read_csv(url, header=None, sep=r"\s+")
    except:
        try:
            return pd.read_csv("housing.csv", header=None, sep=r"\s+")
        except:
            raise Exception("Could not load data")

housing = load_soccer_data()
# Load dataset
housing.columns = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

# Define X (predictors) and y (target variable)
X = housing.drop(columns=["MEDV"]).values
y = housing["MEDV"].values

# Normalize X
X = (X - X.mean(axis=0)) / X.std(axis=0)



results = fourtenmodule.run_and_plot_models(X, y, n_components=2)

# Access the best model's information
print(f"Best model: {results['best_model']}")
print(f"Best RMSE: {results['best_metric']}")
best_trace = results['best_trace']

# access all results
all_results_df = results['all_results']
print(all_results_df)


Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.



=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 3.44s

Testing Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 3 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 3.63s

Testing Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.35s

Testing Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.


  → Passed in 2.03s

Testing Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,419
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 1.37s

Testing Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 1 seconds.
The number of samples is too small to check convergence reliably.
Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.93s

Testing Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 50 tune and 50 draw iterations (200 + 200 draws total) took 0 seconds.
The number of samples is too small to check convergence reliably.
Initializing NUTS using jitter+adapt_diag...


  → Passed in 0.88s

=== Running full analysis on passing models ===

Running Bayesian Linear Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 4 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Linear Regression successfully

Running Bayesian Ridge Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [tau, beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 12 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Ridge Regression successfully

Running Bayesian Lasso...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 5 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Lasso successfully

Running Bayesian Robust Regression...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma, nu]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 6 seconds.


Completed Bayesian Robust Regression successfully

Running Bayesian Variational Inference...


Output()

Finished [100%]: Average Loss = 2,419.2
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian Variational Inference successfully

Running Bayesian PCR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.
Initializing NUTS using jitter+adapt_diag...


Completed Bayesian PCR successfully

Running Bayesian ICR...


Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

Sampling 4 chains for 1_000 tune and 4_000 draw iterations (4_000 + 16_000 draws total) took 2 seconds.


Completed Bayesian ICR successfully

=== Final Results (only passing models) ===

Model                          RMSE       Beta Estimates
----------------------------------------------------------------------
Bayesian Lasso                 23.0216    β0:-0.725  β1:0.778  β2:-0.211  β3:0.683  β4:-1.459  β5:2.766  β6:-0.063  β7:-2.479  β8:1.470  β9:-1.114  β10:-1.845  β11:0.784  β12:-3.579
Bayesian Ridge Regression      23.0995    β0:-0.598  β1:0.510  β2:-0.446  β3:0.677  β4:-0.677  β5:2.497  β6:-0.202  β7:-1.136  β8:0.335  β9:-0.524  β10:-1.440  β11:0.674  β12:-2.489
Bayesian Variational Inference 23.1646    β0:-0.633  β1:0.493  β2:-0.627  β3:0.830  β4:-0.525  β5:2.044  β6:-0.284  β7:-0.816  β8:0.115  β9:-0.442  β10:-1.450  β11:0.677  β12:-1.957
Bayesian Linear Regression     23.2188    β0:-0.510  β1:0.410  β2:-0.500  β3:0.574  β4:-0.443  β5:1.967  β6:-0.263  β7:-0.474  β8:-0.060  β9:-0.461  β10:-1.132  β11:0.552  β12:-1.809
Bayesian Robust Regression     23.2258    β0:-0.522  β1:0.402

KeyError: 'best_model'

## Soccer Players' Market Value
Many redundant variables, expect lasso or ridge to perform best

In [ ]:
# Load dataset (with SSL fix)
def load_soccer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/main/transfermarkt201718.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("transfermarkt201718.csv", sep=";")
        except:
            raise Exception("Could not load data")

soccer = load_soccer_data()

# Data preprocessing pipeline with feature selection
def preprocess_soccer_data(df, response, top_n_features=None):

    # Separate features and target
    y = df[response]
    X = df.drop(columns=[response])
    if top_n_features == None:
        top_n_features = len(X)
    
    # Identify numeric and categorical columns
    numeric_cols = X.select_dtypes(include=np.number).columns.tolist()
    categorical_cols = X.select_dtypes(exclude=np.number).columns.tolist()
    
    # Create preprocessing pipelines
    numeric_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='median')),
        ('scaler', StandardScaler())
    ])
    
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
    ])
    
    # Combine preprocessing steps
    preprocessor = ColumnTransformer(
        transformers=[
            ('num', numeric_transformer, numeric_cols),
            ('cat', categorical_transformer, categorical_cols)
        ])
    
    # Apply preprocessing
    X_processed = preprocessor.fit_transform(X)
    
    # Get feature names
    numeric_features = numeric_cols
    if len(categorical_cols) > 0:
        cat_encoder = preprocessor.named_transformers_['cat'].named_steps['onehot']
        categorical_features = cat_encoder.get_feature_names_out(categorical_cols)
        all_features = numeric_features + list(categorical_features)
    else:
        all_features = numeric_features
    
    # Convert to DataFrame
    X_df = pd.DataFrame(X_processed, columns=all_features)
    
    # Select top N features by variance
    if len(all_features) > top_n_features:
        selector = VarianceThreshold()
        selector.fit(X_df)
        variances = selector.variances_
        
        # Get indices of top N features
        top_indices = np.argsort(variances)[-top_n_features:]
        selected_features = [all_features[i] for i in top_indices]
        
        print(f"\nSelected top {top_n_features} features from {len(all_features)} total features")
        print("Top 10 features by variance:")
        for feat in selected_features[:10]:
            print(f"- {feat}")
        
        return X_df[selected_features], y
    else:
        return X_df, y
print(len(soccer))
# Preprocess data and select top 100 features
X, y = preprocess_soccer_data(soccer, "value")


# Now run models
results = fourtenmodule.run_and_plot_models(X, y)

# Access results
print(f"\nBest performing model: {results['best_model']}")
print(f"Best RMSE: {results['best_metric']:.4f}")

print("\nModel Comparison Results:")
print(results['all_results'])

Only 50 samples per chain. Reliable r-hat and ESS diagnostics require longer chains for accurate estimate.


2232

Selected top 2232 features from 2837 total features
Top 10 features by variance:
- player_Jonjoe Kenny
- player_Kevin Akpoguma
- player_Gerson
- player_Cuco Martina
- nationality_cw CUW
- player_Benjamin Hübner
- player_Alfred Duncan
- player_Kevin Mirallas
- player_Oumar Niasse
- player_Jordan Pickford

=== Running timing tests (max 5s per model) ===

Testing Bayesian Linear Regression...


Initializing NUTS using jitter+adapt_diag...
Multiprocess sampling (4 chains in 4 jobs)
NUTS: [beta, sigma]


Output()

### Cancer Survival Data
Highly collinear

In [ ]:
def load_cancer_data():
    try:
        url = "https://raw.githubusercontent.com/javmencia/STA410/refs/heads/main/METABRIC_RNA_Mutation.csv"
        return pd.read_csv(url, sep=";")
    except:
        try:
            return pd.read_csv("housing.csv", header=None, sep=r"\s+")
        except:
            raise Exception("Could not load data")

housing = load_soccer_data()
# Load dataset
housing.columns = ["CRIM", "ZN", "INDUS", "CHAS", "NOX", "RM", "AGE", "DIS", "RAD", "TAX", "PTRATIO", "B", "LSTAT", "MEDV"]

# Define X (predictors) and y (target variable)
X = housing.drop(columns=["MEDV"]).values
y = housing["MEDV"].values

# Normalize X
X = (X - X.mean(axis=0)) / X.std(axis=0)



results = fourtenmodule.run_and_plot_models(X, y, n_components=2)

# Access the best model's information
print(f"Best model: {results['best_model']}")
print(f"Best RMSE: {results['best_metric']}")
best_trace = results['best_trace']

# Or access all results
all_results_df = results['all_results']


In [ ]:
print(results['all_results'])

